cpp/src/community/k_truss_impl.cuh

```diff
@@ -456,6 +204,537 @@ struct generate_p_r_or_q_r_from_p_q {
+
+  auto vertex_pair_buffer_unique = allocate_dataframe_buffer<thrust::tuple<vertex_t, vertex_t>>(
+    unique_pair_count, handle.get_stream());
+
+  thrust::reduce_by_key(handle.get_thrust_policy(),
+                        vertex_pair_buffer_begin,
+                        vertex_pair_buffer_begin + vertex_pair_buffer_src.size(),
+                        decrease_count_tmp.begin(),
+                        get_dataframe_buffer_begin(vertex_pair_buffer_unique),
+                        decrease_count.begin(),
+                        thrust::equal_to<thrust::tuple<vertex_t, vertex_t>>{});
+
+  cugraph::edge_bucket_t<vertex_t, void, true, multi_gpu, true> edges_to_decrement_count(handle);
+  edges_to_decrement_count.insert(std::get<0>(vertex_pair_buffer_unique).begin(),
+                                  std::get<0>(vertex_pair_buffer_unique).end(),
+                                  std::get<1>(vertex_pair_buffer_unique).begin());
+
+  cugraph::transform_e(
+    handle,
+    cur_graph_view,
+    edges_to_decrement_count,
+    cugraph::edge_src_dummy_property_t{}.view(),
+    cugraph::edge_dst_dummy_property_t{}.view(),
+    e_property_triangle_count.view(),
+    [vertex_pair_buffer_begin = get_dataframe_buffer_begin(vertex_pair_buffer_unique),
+     vertex_pair_buffer_end   = get_dataframe_buffer_end(vertex_pair_buffer_unique),
+     decrease_count           = decrease_count.data()] __device__(auto src,
+                                                        auto dst,
+                                                        thrust::nullopt_t,
+                                                        thrust::nullopt_t,
+                                                        edge_t count) {
+      auto e = thrust::make_tuple(src, dst);
+      auto itr_pair =
+        thrust::lower_bound(thrust::seq, vertex_pair_buffer_begin, vertex_pair_buffer_end, e);
+
+      if ((itr_pair != vertex_pair_buffer_end) && (*itr_pair == e)) {
+        auto idx_pair = thrust::distance(vertex_pair_buffer_begin, itr_pair);
+        return count - decrease_count[idx_pair];
+      }
+
+      return count;
+    },
+    e_property_triangle_count.mutable_view(),
+    true);
+};
+
+template <typename vertex_t,
+          typename edge_t,
+          typename weight_t,
+          typename optional_graph_view_t,
+          bool multi_gpu,
+          bool is_q_r_edge,
+          bool is_p_q_edge>
+void find_unroll_p_q_q_r_edges(
+  raft::handle_t const& handle,
+  graph_view_t<vertex_t, edge_t, false, multi_gpu>& cur_graph_view,
+  optional_graph_view_t const& graph_q_r,
+  edge_property_t<graph_view_t<vertex_t, edge_t, false, multi_gpu>, edge_t>&
+    e_property_triangle_count,
+  edge_property_t<graph_view_t<vertex_t, edge_t, false, multi_gpu>, bool>& tmp_edge_mask,
+  raft::device_span<vertex_t> weak_edgelist_srcs,
+  raft::device_span<vertex_t> weak_edgelist_dsts,
+  std::optional<rmm::device_uvector<vertex_t>> renumber_map,
+  bool do_expensive_check)
+{
```

Comment:
I think this function is too busy. This code tries to handle both p,q & q,r unrolling in one function, but this makes the code harder to follow.

Reply:
Let's try to decompose this function.

Line 386 - Line 467, given a graph view object and invalid edges, this function accumulates triangles. This can be one function and when we are dealing with (p, q) edges, we can pass graph_view and p,q edges. When we are dealing with (q,r) edges, we can create another graph view object (i.e. graph_q_r) and q,r edges. One function doing two different tasks can be quite confusing. One function just doing one work (but returns different results based on input arguments) is much easier to follow & maintain.

Reply:
I am having hard time understanding the code after this.

I guess this code updates triangle counts from the accumulated triangles (and we need to avoid overcompensation if q,r edges).

This can be another function taking a graph view object, edge_triangle_counts, accumulated triangles, and optional additional information to avoid overcompensation (this optional data can bet std::nullopt for (p,q) edges). And I am not sure why this code un-renumbers here. Let's discuss more about this tomorrow. I think we can simplify this code.